### DISCUSSION TITLE
DJL v0.14.0 release

### DISCUSSION THREAD

**Discussion by frankfliu** [deepjavalibrary/djl#1360] [News]
DJL v0.14.0 updates the engines [PyTorch](https://github.com/awslabs/djl/tree/master/engines/pytorch/pytorch-engine) to 1.9.1 and introduces several new features:

## Key Features

* Upgrades [PyTorch](https://github.com/awslabs/djl/tree/master/engines/pytorch/pytorch-engine) engine to 1.9.1
* Adds support for [Neuron SDK 1.16.1](https://awsdocs-neuron.readthedocs-hosted.com/en/latest/release-notes/index.html#latest-neuron-release)
* Adds autoscale in [djl-serving](https://github.com/deepjavalibrary/djl-serving) for AWS Inferentia model
* Introduces [OpenCV extension](https://github.com/deepjavalibrary/djl/tree/master/extensions/opencv) to provide high performance image processing
* Adds support for older version of PyTorch engine, user now can use PyTorch 1.8.1 with latest DJL
* Adds support for precxx11 PyTorch native library in auto detection mode
* Adds AWS Inferentia support in [djl-bench](https://snapcraft.io/djlbench)
* Adds support for [TorchServe](https://github.com/pytorch/serve) .mar format, user can deploy TorchServe model archive in djl-serving

## Enhancement

* Introduces several new features in djl-serving:
    * Adds autoscale feature for AWS Inferentia ([#31](https://github.com/deepjavalibrary/djl-serving/pull/31))
    * Creates SageMaker hosting compatible docker image for AWS Inferentia ([#36](https://github.com/deepjavalibrary/djl-serving/pull/36))
    * Adds auto detect number of neuron cores feature for AWS Inferentia ([#34](https://github.com/deepjavalibrary/djl-serving/pull/34))
    * Adds autoscale support for SageMaker style .tar.gz model ([#35](https://github.com/deepjavalibrary/djl-serving/pull/35))
    * Adds support to load torchserve model ([#32](https://github.com/deepjavalibrary/djl-serving/pull/32))
    * Adds support to pip installed dependency per model ([#37](https://github.com/deepjavalibrary/djl-serving/pull/37))
    * Adds custom environment variable support for python engine ([#29](https://github.com/deepjavalibrary/djl-serving/pull/29))
    * Adds nested folder support in model archive file ([#38](https://github.com/deepjavalibrary/djl-serving/pull/38))
    * Improves model status with model version support ([#25](https://github.com/deepjavalibrary/djl-serving/pull/25))
    * Adds model warn up feature for python engine. ([#23](https://github.com/deepjavalibrary/djl-serving/pull/23))
    * Adds WorkLoadManager.unregisterModel ([#33](https://github.com/deepjavalibrary/djl-serving/pull/33))
    * Adds testing tool to test python model locally ([#22](https://github.com/deepjavalibrary/djl-serving/pull/22))
    * Adds set python executable path for python engine ([#21](https://github.com/deepjavalibrary/djl-serving/pull/21))
    * Creates Workflow for ModelServing ([#26](https://github.com/deepjavalibrary/djl-serving/pull/26))
* Adds OpenCV extension ([#1331](https://github.com/deepjavalibrary/djl/pull/1331))
* Introduces several new features in djl-bench:
    * Adds support for AWS Inferentia ([#1329](https://github.com/deepjavalibrary/djl/pull/1329))
* Introduces several new features in Apache MXNet engine:
    * Implements LayerNorm for Apache MXNet ([#1342](https://github.com/deepjavalibrary/djl/pull/1342))
* Introduces several new features in PyTorch engine:
    * Upgrades PyTorch to 1.9.1 ([#1297](https://github.com/deepjavalibrary/djl/pull/1297))
    * Implements padding to bert tokenizer ([#1328](https://github.com/deepjavalibrary/djl/pull/1328))
    * Makes pytorch-native-auto package optional ([#1326](https://github.com/deepjavalibrary/djl/pull/1326))
    * Adds support to use different version of PyTorch native library ([#1323](https://github.com/deepjavalibrary/djl/pull/1323))
    * Adds map_location support for load model from InputStream ([#1314](https://github.com/deepjavalibrary/djl/pull/1314))
    * Makes map_location optional ([#1312](https://github.com/deepjavalibrary/djl/pull/1312))
* Introduces several new features in TensorFlow Lite engine:
    * Makes tensor-native-auto package optional ([#1301](https://github.com/deepjavalibrary/djl/pull/1301))
* Introduces several API improvements:
    * Adds support for nested folder in model archive ([#1349](https://github.com/deepjavalibrary/djl/pull/1349))
    * Improves translator output error message ([#1348](https://github.com/deepjavalibrary/djl/pull/1348))
    * Improves Predictor API to support predict with device ([#1346](https://github.com/deepjavalibrary/djl/pull/1345))
    * Improves BufferedImageFactory.fromNDArray performance ([#1339](https://github.com/deepjavalibrary/djl/pull/1339))
    * Adds support for downloading .mar file ([#1338](https://github.com/deepjavalibrary/djl/pull/1338))
    * Adds debugging toString to Input and Output ([#1327](https://github.com/deepjavalibrary/djl/pull/1327))
    * Refactors BERT Translator and Tokenizer ([#1318](https://github.com/deepjavalibrary/djl/pull/1318))
    * Makes question answering model serving ready ([#1311](https://github.com/deepjavalibrary/djl/pull/1311))
    * Refactors minMaxWorkers from ModelInfo to WorkerPool ([#30](https://github.com/deepjavalibrary/djl-serving/pull/30))

## Documentation and examples

* Adds huggingface Inferentia serving example ([#184](https://github.com/deepjavalibrary/djl-demo/pull/184))
* Adds [AWS SageMaker hosting document](https://github.com/deepjavalibrary/djl-demo/blob/master/huggingface/inferentia/deploy_on_sagemaker.md)
* Adds python hybrid engine demo ([#182](https://github.com/deepjavalibrary/djl-demo/pull/182))
* Improves DJL examples project gradle build script ([#1344](https://github.com/deepjavalibrary/djl/pull/1344))

## Breaking change

* PyTorch 1.9.1 no longer supports Amazon Linux 2, AL2 user has to use pytorch-native-cpu-precxx11
* Image.Type is removed and Image.duplicate() function no longer take Image.Type as input
* Image.getSubimage() is renamed to Image.getSubImage()
* PaddlePaddle model loading may break due to prefix changes.

## Bug Fixes

* Fixes 2nd inference throw exception bug ([#1351](https://github.com/deepjavalibrary/djl/pull/1351))
* Fixes calculation for SigmoidBinaryCrossEntropyLoss from sigmoid ([#1345](https://github.com/deepjavalibrary/djl/pull/1345))
* Fixes jar model url download bug ([#1336](https://github.com/deepjavalibrary/djl/pull/1336))
* Fixes memory in Trainer.checkGradients ([#1319](https://github.com/deepjavalibrary/djl/pull/1319))
* Fixes NDManager is closed bug ([#1308](https://github.com/deepjavalibrary/djl/pull/1308))
* Fixes PyTorch GPU model loading issue ([#1302](https://github.com/deepjavalibrary/djl/pull/1302))
* Fixes MXNet EngineException message ([#1300](https://github.com/deepjavalibrary/djl/pull/1300))
* Fixes python resnet18 demo model GPU bug ([#24](https://github.com/deepjavalibrary/djl-serving/pull/24))
* Fixes python engine get_as_bytes() bug ([#20](https://github.com/deepjavalibrary/djl-serving/pull/20))

## Contributors

This release is thanks to the following contributors:

* Frank Liu ([@frankfliu](https://github.com/frankfliu))
* Jake Lee ([@stu1130](https://github.com/stu1130))
* enpasos ([@enpasos](https://github.com/enpasos))
* Qing Lan ([@lanking520](https://github.com/lanking520))
* Zach Kimberg ([@zachgk](https://github.com/zachgk))

<hr /><em>This discussion was created from the release <a href='https://github.com/deepjavalibrary/djl/releases/tag/v0.14.0'>DJL v0.14.0 release</a>.</em>
